### DISCUSSION TITLE
Bei jedem Firmwareupdate Probleme!

### DISCUSSION THREAD

**Discussion by Snusme83** [hoylabs/OpenDTU-OnBattery#1186] [General]
![Funktioniert wieder nicht](https://github.com/user-attachments/assets/692ea0c5-fb9d-4c87-9e45-0d0cda420298)

Servus @ all, hat mal jemand ne vernünftige Anleitung, wie man auf das aktuellste Release updaten kann?

Hatte die vom Juni drauf und ich Trottel hab es wieder mal riskiert upzudaten.

Habe ein ESP32-S3 mit 8MB und mir gelingt das Update schon wieder nicht. Nicht mal über das Webinstall!

**Comment by Akhilananda** [comment]
Bei opendtu fusion 2.1 funktioniert nur die Firmware 2024.01.26 einwandfrei alle anderen nachfolgenden Versionen sind fehlerhaft :-(
Da ich nur ein Anwender und kein Mitentwickler der Software bin stehe ich ziemlich im Regen. 
Ich mache jetzt einfach kein update mehr und bleibe bei der  Firmware 2024.01.26

**Reply by Manos1966** [reply]
> Bei opendtu fusion 2.1 funktioniert nur die Firmware 2024.01.26 einwandfrei alle anderen nachfolgenden Versionen sind fehlerhaft :-(

Das kann einfach nicht der Fall sein.
Ich habe selber zwar keine, die opendtu Fusion ist aber die meist gekauft (fertige) Platine und hat exzellenten Service. 
Ich kann mir nicht vorstellen, dass seit 6 Monate keine passende Firmware drauf funktioniert 😉

**Reply by Snusme83** [reply]
Die vom Juni hat perfekt funktioniert! Ich könnte mich so ärgern, jetzt stehe ich wieder stundenlange in der Garage um die Firmware auf den Mikrocontroller zu bekommen. Wenn ich es diesmal wieder schaffe werde ich nie wieder updaten!

**Reply by schlimmchen** [reply]
> Bei opendtu fusion 2.1 funktioniert nur die Firmware 2024.01.26 einwandfrei alle anderen nachfolgenden Versionen sind fehlerhaft :-(

Da du außer [einer merkwüdigen Diskussion](https://github.com/search?q=repo%3Ahelgeerbe%2FOpenDTU-OnBattery+author%3AAkhilananda&type=discussions) kein Issue und auch sonst nichts von dir hast hören lassen, kann niemand wissen, dass du dein Fusion nicht in Gang bringen kannst. An der Firmware liegt es jedenfalls nicht.

**Reply by AndreasBoehm** [reply]
Ich hab mittlerweile drei Fusion Boards im Einsatz, meistens auf dem aktuellen Entwicklungsstand und habe keinerlei Probleme.

**Reply by Akhilananda** [reply]
Ich habe bereits an anderer Stelle über das Problem ausführlich berichtet. 
Was solls mit der Firmware 2024.01.26 kann ich gut leben 
Ich darf mich nur nicht zu updates verführen lassen.
Firmware 2024.01.26 zeigt in der Konsole auch DPL Daten an.
Firmware 2024.08.18 zeigt in der Konsole keine DPL Daten an
Neue opendtu Fusion kaufen? Nein.

**Reply by schlimmchen** [reply]
> Ich habe bereits an anderer Stelle über das Problem ausführlich berichtet.

Link?

> Firmware 2024.08.18 zeigt in der Konsole keine DPL Daten an

Dann hast du entweder "verbose logging" ausgeschaltet (wobei das vermutlich schon im Januar implementiert war) oder du schaust auf die serielle Konsole und hast die falsche Firmware Variante installiert (_usb versus ohne _usb).

> Neue opendtu Fusion kaufen? Nein.

Natürlich nicht, das Board kann ja nichts dafür...

**Comment by Manos1966** [comment]
Wiki probiert?
https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/WebFlasher_Introduction

**Reply by Snusme83** [reply]
Nach dieser Anleitung hab ich es ja gemacht!

**Reply by Snusme83** [reply]
Muss der Mikrocontroller eine gewisse Zeit stromlos sein (sprich nicht am USB?) Muss ich sämtliche Pins abstecken?

**Reply by Manos1966** [reply]
Wie gut ist dein English? Verstehst du die Fehler-Meldung?
Hochstwahrscheinlichj muss du:
Before starting the flash process you have to manually put the ESP into bootloader mode. (Most boards are able to do this automatically, but your board is not):

PC neu starten (man weiss nie...)
ESP-S3 angeschlossen HW (falls) entfernen
BOOT knopf gedruckt halten und ESP-S3 am PC anschliessen
BOOT loss lassen
Webflasher starten

**Reply by Snusme83** [reply]
Wenn ich den Boot- Button auf dem Mikrocontroller gedrückt halte bringt das gar nix!

Gibt es da ne bestimmte Kombination? GPIO0 und GND zu verbinden hilft auch nix!

**Reply by Manos1966** [reply]
> Muss ich sämtliche Pins abstecken?

Gut erraten! Die richtige Antwort sollte "Nein" sein, aber man weiss nie welche PINs du angeschlossen hast... manche davon koennen einen Einfluss haben!

**Reply by Manos1966** [reply]
> GPIO0 und GND zu verbinden hilft auch nix!

GPIO0 und GND ist BOOT-Knopf

Deine ESP32-S3 hat zwei Typ-C Anschlusse. Welchen benutzt du?

**Reply by spcqike** [reply]
> Gibt es da ne bestimmte Kombination? 

ja. Gpio0 und GND verbinden bevor man Spannung anlegt.

**Reply by schlimmchen** [reply]
> Deine ESP32-S3 hat zwei Typ-C Anschlusse. Welchen benutzt du?

Ich habe scheinbar keinen Erfolg mit dem Port, der über die USB-UART-Bridge mit dem ESP32 verbunden ist. Ein China-CH343 ist das. Es funktioniert aber problemlos mit esptool.py von der Konsole. Hast du ähnliche Erfahrungen gemacht?

**Reply by Snusme83** [reply]
Danke schon mal an alle, die sich meinem Problem angenommen haben.

Besonderer Dank in diesem Fall an Manos1966 und dem Zufall dass ich scheinbar die richtigen PINs abgezogen hab.

Tatsächlich hat es scheinbar zu tun mit dem angesteckten Pylontech(s) (4 PINS).

Nachdem ich diese abgezogen hab, konnte ich per webflash die aktuellste Firmware installieren und das erfolgreich. Die letzten Male habe ich dann immer einen neuen Mikrokontroller verwendet und deswegen funktionierte es dann immer, weil ich sämtliche Kabel dann neu angesteckt habe!

![Passt wieder](https://github.com/user-attachments/assets/e176569f-b9e3-4ae0-8082-8ad005db1e62)

Das sind im Übrigen meine Dashboards, welche ich in Grafana gebastelt hab. Sie funktionieren auch wieder!
![Grafana Dashboard 1](https://github.com/user-attachments/assets/33fddb0c-73a1-4dcc-b277-64419fb65872)
![Grafana Dashboard 2](https://github.com/user-attachments/assets/fda2ba2c-e9d5-407b-879b-581c9c00706e)

Danke an die tolle Community hier!

**Reply by Manos1966** [reply]
> und dem Zufall dass ich scheinbar die richtigen PINs abgezogen hab.

Frag nie woher ich das wusste.... 🙊 

🤣 🤣 🤣 🤣

**Reply by Snusme83** [reply]
Ich denke du standest auch schon eine gewisse Zeit in der Garage, wie ich! :-)

**Comment by schlimmchen** [comment]
> Ich könnte mich so ärgern, jetzt stehe ich wieder stundenlange in der Garage um die Firmware auf den Mikrocontroller zu bekommen.

@Snusme83 Es ist verständlich, dass das ärgerlich ist. Ich kenne das Gefühl. Aber es ist nicht schwer.

Welche Hardware hast du denn im Detail? Hat das Ding einen Button, der mit "Boot" beschriftet ist? Wenn man den drückt und gedrückt hält während der ESP32 startet, dann landet dieser im bootloader. Der Button schließt GPIO0 mit GND kurz, wie du schon erwähnt hast. Also am besten stromlos machen, den Boot-Knopf drücken und gedrückt halten, währenddessen das USB-Kabel einstecken, erst dann den Button loslassen. Dann den Web-Flasher starten.

Hast du den richtigen Port ausgewählt im WebFlasher?

**Reply by Snusme83** [reply]
Hi schlimmchen, ja das habe ich alles so gemacht - ich bin ja nicht doof und dem Englischen bin ich auch nicht ganz unfähig. Zudem gibt es den Google Übersetzer. Hab jetzt mal die 4 Pins des Mikrocontrollers befreit, welche mit dem Pylontech verbunden sind.

Siehe da die Firmware wird per webinstall installiert und war scheinbar erfolgreich! Ich werde noch berichten!

**Reply by schlimmchen** [reply]
Sehr gut! Das ist allerdings interessant. Lass uns bitte wissen, welche Pins du da verwendet hast. Teile vielleicht einfach deine funktionierende pin_mapping.json, dann können wir das ableiten.

Achso... Versorgt da vielleicht irgendetwas rückwärts das Board mit Spannung? Dann würde der ESP32 nicht neu starten, und Boot-Knopf halten bringt dann nichts. Egal ob das bei dir so ist oder nicht, das wäre gut allen Benutzern nahezulegen, sie mögen stattdessen den Reset-Knopf des Boards nutzen, dann wird dieses Problem obsolet.

**Reply by Snusme83** [reply]
Also, ich habe folgenden ESP32-S3:
![ESP32-S3GPIO](https://github.com/user-attachments/assets/a5cefce2-c531-46a8-af36-7a9cd9cbffb2)

Pylontech ist mit dem Mikrocontroller mit 4 Pins verbunden:

![Pylontechbelegung](https://github.com/user-attachments/assets/4eb0032c-b89c-47b6-a252-4647c1929f0c)

Recht herzlichen Dank nochmal. Ihr seid die Besten. Dachte schon ich verbringe die Nacht in der Garage!

**Reply by Manos1966** [reply]
Machen wir eine Lotto-Gemeinschaft? 🤣  Von 38 PINs musstest du diese zwei auswaehlen?

Ja, nicht die beste Wahl...

Die Info haben wir aber richtig gut versteckt... nur fuer Kenner 🤐 
![S3 Misery PINs](https://github.com/user-attachments/assets/ca093334-e825-4842-bd0d-486dd3d16c11)

**Reply by Snusme83** [reply]
Hä? Also scheinbar habe ich Pins ausgewählt die suboptimal sind. Allerdings haben 941 User die aktuellste OpenDTUonBattery heruntergeladen und scheinbar hatte keiner Probleme bei dem seriellen Update!

**Reply by Snusme83** [reply]
Im Übrigen verwalte ich eine Lottogemeinschaft seit mittlerweile zwei Jahrzehnten! :-)

**Reply by Manos1966** [reply]
Ich habe das Wiki upgedated (geupdated?) 
Beim flashen sollte jegliche angeschlossene HW entfernt sein.
Es spart viel Frust und Aerger.

**Reply by Snusme83** [reply]
Also für TX und RX vom Pylontech soll ich ander Pins wählen? Welche denn? Gut, du weißt jetzt meine komplette Belegung nicht, aber ich bin der Typ, der drei Victron SmartSolar 100/20 am Controller hängen hat.

**Reply by schlimmchen** [reply]
> Beim flashen sollte jegliche angeschlossene HW entfernt sein.

Also die Leute nervös zu machen ist auch nicht sinnvoll. Das Problem ist doch verstanden und es handelt sich nur um die Wahl von UART0 für die Verbindung zu einem Peripheriegerät, was hier Probleme gemacht hat.

> Also für TX und RX vom Pylontech soll ich ander Pins wählen?

Meiner Meinung nach nicht. Das ist vollkommen in Ordnung, dort etwas anzuschließen. Du musst dann allerdings den nativen USB-Port an deinem ESP32-S3 Board benutzen um Firmwareupdates zu machen. Dann gibt es auch kein Problem. Du musst vorhin den Port an der UART-USB-Bridge verwendet haben, und der macht dann natürlich Probleme, wenn an UART0 was dranhängt.

**Reply by Snusme83** [reply]
> Ich habe das Wiki upgedated (geupdated?) Beim flashen sollte jegliche angeschlossene HW entfernt sein. Es spart viel Frust und Aerger.

Ich habe im Wiki geschaut. Konnte aber keine geupdatete/ upgedatete Textpassage finden!?!

**Reply by Manos1966** [reply]
> Ich habe im Wiki geschaut. Konnte aber keine geupdatete/ upgedatete Textpassage finden!?!

"We agree to disagree" 😜 
https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/WebFlasher_Introduction

![note First installation](https://github.com/user-attachments/assets/08d2ed72-7bdf-4985-a1ed-c486acdf66b7)

**Comment by Manos1966** [comment]
> Ich habe scheinbar keinen Erfolg mit dem Port, der über die USB-UART-Bridge mit dem ESP32 verbunden ist.

genau meine Erfahrung auch!

**Reply by schlimmchen** [reply]
Ich muss im "Preparing Installation" Schritt (noch einmal) manuell einen Reset in den Bootloader machen, dann geht's. Alternativ kann man den BOOT Button auch halten, bis der Flash gelöscht wird. Das muss damit zu tun haben, dass der WebFlasher einen Reset auslöst, aber der ESP startet dann normal statt in den Bootloader.

**Reply by Snusme83** [reply]
Da bin ich raus. Bin kein Informatiker! Aber ich werde mir die Infos morgen nochmal genauer anschauen und darüber grübeln, aber meistens liegt es daran, dass mir bei euren Begriffen einfach das Wissen fehlt, da ich Laie bin.

**Reply by Nailuj1** [reply]
> Ich muss im "Preparing Installation" Schritt (noch einmal) manuell einen Reset in den Bootloader machen, dann geht's. Alternativ kann man den BOOT Button auch halten, bis der Flash gelöscht wird. Das muss damit zu tun haben, dass der WebFlasher einen Reset auslöst, aber der ESP startet dann normal statt in den Bootloader.

Was heißt das, wenn ich die aktuelle FW auf meinen Fusion 2.2 flashen will?

**Reply by schlimmchen** [reply]
Das heißt dass du bitte versuchst der doku https://opendtu-onbattery.net/firmware/howto/upgrade_8mb/ zu folgen und Rückmeldung gibst, wenn da etwas fehlt oder du trotzdem Probleme hast 😉

**Comment by Snusme83** [comment]
@schlimmchen: Das ist mein funktionierendes Pinmapping!
![Pinmapping 1](https://github.com/user-attachments/assets/bec64875-614f-4ea7-b002-790c6908b592)
![Pinmapping 2](https://github.com/user-attachments/assets/433d349d-b4a6-4689-a504-4675a30fb544)
![Pinmapping 3](https://github.com/user-attachments/assets/045f595d-7161-462c-816a-8210ba5b001f)

**Comment by Snusme83** [comment]
Das "Problem" dieser Community ist eigentlich nur, dass ihr es stetig schafft die PERFEKTE OpenDTUonBattery immer wieder zu verbessern. Ihr geht auf sämtliche Thematiken ein. Das Battery- Symbol abgesetzt vom Text, ...! Mir gefällt diese neue Hobby, aber ich wäre mal froh, wenn ich ein Update machen könnte, dies erfolgreich ist und ich mich dem gemütlichen Teil des Abends hingeben könnte.

Scheinbar wird das Problem der Firmwareaktualisierung bei mir dadurch verursacht, weil ich falsche Pins des Mikrocontrollers verwende, bzw. diese beim Upgrade angesteckt sind!

**Reply by schlimmchen** [reply]
Jo. Also ich wollte ja auch gerne was anderes machen, aber ich bohre stattdessen die Doku auf und probiere selbst noch ein wenig herum, was so die Stolpersteine sind. Ich brauch das alles eigentlich nicht, ich weiß mir zu helfen, aber für jemand der nur schnell ein Update machen will, soll das natürlich gut uns hinreichend erklärt sein, wie sich zu helfen ist.

Für die Zukunft, du bist ja jetzt auf dem neuen Partitionslayout, kommst du ja mit [OTA updates](http://127.0.0.1:8000/firmware/update/#using-flash-tool) aus. Da kannst du eigentlich nur den einen Fehler machen: factory.bin Datei hochladen statt die ohne factory im Namen. Es sollte für dich lange Zeit keinen Grund geben, per Kabel aktualisieren zu müssen.

**Reply by Snusme83** [reply]
Das hab ich schon mitbekommen, dass zukünftige Updates mit OTA- Updates möglich sein werden. 

Brauche ich dann:


opendtu-onbattery-generic_esp32s3_usb.factory.bin                       oder

opendtu-onbattery-generic_esp32s3_usb.bin?

**Reply by schlimmchen** [reply]
Naja, komm schon, das hab ich sowohl in meiner Nachricht erklärt als auch ist es im verlinkten Doku-Artikel beschrieben. Das kaue ich dir jetzt nicht vor. Schau dir den [Artikel](https://opendtu-onbattery.net/firmware/update/) an.

Achso, mist, der Link ist kaputt. Habe ich korrigiert.

**Reply by Snusme83** [reply]
Sorry, ich muss mir das morgen alles noch mal in Ruhe reinziehen. Das Problem ist dass ihr/ du von einem gewissen Know-How ausgeht, das der Otto-Normal-Verbraucher aber nicht hat (Problem liegt beim "Laien")!

Ihr macht echt ein klasse Arbeit, welche mit einem brutalen zeitlichen Aufwand verbunden ist. Aber wie gesagt ich finde euer Projekt BOMBE. Danke, dass ihr soviel Geduld und Nerven habt, dieses Projekt stets voranzutreiben und zu "verbessern", obwohl es perfekt ist.

Hut ab, kann ich da nur sagen!

**Reply by Snusme83** [reply]
@schlimmchen: Da ich ständig Probleme mit dem Flashen des ESP-32 S3 hatte, habe ich mir folgendes Breakout-Board angeschafft:

![Breakout Board](https://github.com/user-attachments/assets/9742673c-2cdc-41db-992c-eb66de5e96a5)

Jetzt brauche ich nicht jedes Mal die Verkabelung lösen vom Microcontroller, sondern ich ziehe den Microcontroller vom Breakout- Board und kann ohne Verkabelung Flashen bzw. sollte irgendwann mal was mit dem Microcontroller ned stimmen, kommt einer der Reserve Microcontroller ins Breakout- Board und die OpenDTUonBattery läuft wieder! :-)

Für dich nichts Neues, aber ich glaub das war ein sinnvolles Investment für mich!

Die OpenDTUonBattery ist Wahnsinn. Großen Dank und Respekt, was ihr da auf die Füße gestellt habt und der Allgemeinheit zur Verfügung stellt - Hut ab. Danke auch, für die Zeit, die ihr opfert, um "Laien" wie mir zu helfen.

**Reply by Manos1966** [reply]
Ich mache das gleiche.
Falls auch andere Mitglieder ueber ESP32-S3 BreakoutBoards denken, die gibt es zur Zeit sehr günstig 👍 

![S3 Breakoutboard](https://github.com/user-attachments/assets/532b211b-1b49-46bd-8318-5caec765129e)

**Reply by schlimmchen** [reply]
> Danke auch, für die Zeit, die ihr opfert, um "Laien" wie mir zu helfen.

Naja, "opfern"... Ich seh' das eher als Investment :wink:

> Jetzt brauche ich nicht jedes Mal die Verkabelung lösen vom Microcontroller

Das ist definitiv geschickt :muscle: Allerdings darf ich dazu sagen, dass auch wenn dein Board super hübsch aussieht... Das was Manos zeigt ist sicherlich besser, weil die Schraubklemmen den Steckverbindern überlegen sind... :man_shrugging:

**Comment by lindingerm** [comment]
Ich hatte auch Probleme mit dem Firmwareupdate und möchte hier meinen Lösungsweg beschreiben.

Um den "OpenDTU Fusion" mit dem "OpenDTU-OnBattery - Community WebFlasher" zum flashen zu bringen musste ich die USB Verbindung trennen danach die BOOT Taste gedrückt halten und wieder anstecken und erst dann die BOOT Taste loslassen.

Danach hat der WebFlasher zwar die aktuelle Firmware geflasht und anscheinend erfolgreich mit der Meldung "Installation complete!" beendet aber nach einem Neustart war dann noch immer die alte Firmware drauf.

Die Lösung war dann im WebFlasher das ganze zu wiederholen aber nicht "Up/Downgrade" sondern "Factory Reset" auszuwählen. Da ist dann halt alles neu zu konfigurieren oder das Backup der config und pin_mapping einzuspielen.

So ein schlechtes Update Erlebnis hatte ich noch bei keinem Gerät aber nun läufts.

**Reply by schlimmchen** [reply]
> Um den "OpenDTU Fusion" mit dem "OpenDTU-OnBattery - Community WebFlasher" zum flashen zu bringen musste ich die USB Verbindung trennen danach die BOOT Taste gedrückt halten und wieder anstecken und erst dann die BOOT Taste loslassen.

Das ist [in der Doku beschrieben](https://opendtu-onbattery.net/firmware/flash_esp/#bootloader-mode) und von der [entsprechenden Stelle in der Doku über den WebFlasher](https://opendtu-onbattery.net/firmware/webinstall/#troubleshooting) verlinkt.

**Comment by uputer** [comment]
Bedauerlicherweise muss ich mich hier noch einmal anschließen. 

In der Hoffnung, mit euren geballtem Schwarmwissen weiterzukommen.
Ich habe ein Fusion Board von den AllianceApps (ESP32-S3-), und versuche schon den halben Tag das mit der neuen Version zum Laufen zu bringen. Neuestes Release.
Webflasher geht gar nicht, dödelt ewig rum um dann abzubrechen.
ESP -Tool geht (flashed), der ESP startet aber nicht. Boot drücken und starten, RST alles das was ihr hier als Tipps gegeben habt probiert.

**Log:**
ESP-ROM:esp32s3-20210327
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x7 (TG0WDT_SYS_RST),boot:0x8 (SPI_FAST_FLASH_BOOT)
Saved PC:0x40043ac8
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x44c
load:0x403c9700,len:0xbd8
load:0x00000000,len:0x75647261
Invalid image block, can't boot.
ets_main.c 329 

Interessanterweise funktioniert der Flasher von der OpenDTU, der von der oB dreht sich tot ohne etwas zu bewirken.
(Allerdings bootet der auch nicht ......)
Stehe ich da irgendwo auf dem Schlauch?

Meine AHOYDTU kann mir wenigstens die HM anzeigen, aber für die HMS bräuchte ich doch schon die OpenDTU ...

**Reply by Snusme83** [reply]
Servus, kann es sein, dass du noch Kabel an den GPIOs stecken hast. Das war bei mir das Problem!

**Reply by schlimmchen** [reply]
Hast du [Doku](https://opendtu-onbattery.net/firmware/flash_esp/#bootloader-mode) schon auswendig gelernt und befolgt? Zum Webflasher gibt es auch noch ein Abschnitt mit [Troubleshooting](https://opendtu-onbattery.net/firmware/webinstall/) Hinweisen.

**Comment by uputer** [comment]
Danke für Eure Hinweise.

Welche Kabel? An meiner Platine sehe ich keine. Siehe Bild.

Ja, die Doku habe ich durch.
Ich habe mehrere USB Ports am Rechner ausprobiert, mehrere USB Kabel, mehrfach den Trick mit Boot Taste drücken und dann starten etc.

Mit meinen Tasmotas habe ich solche Flash Probleme noch nicht gehabt.

Daher bin ich auch so ratlos...

Danke.
![IMG_20250129_094724_546](https://github.com/user-attachments/assets/c6bf1cc6-a98e-4726-860e-913315dacd6b)

**Reply by schlimmchen** [reply]
Welchen Webflasher verwendest du? [Diesen](https://solar.metacontrol.eu/opendtu-onbattery-webflasher/)? Ich fürchte dem können wir nicht mehr trauen, weil nicht gepflegt.

Hast du Python auf deinem Rechner? esptool.py würde vermutlich zuverlässig funktionieren.

**Comment by Harisjems** [comment]
Webflasher funktioniert bei mir auch nicht, Flashtool dagegen ohne Probleme mit Factory-Firmware. Am Treiber liegt es nicht. Tasmota Webflasher funktioniert ohne Probleme z. B. mit Factory OpenDTU on Battery Firmware bei FusionBoard.  Bei der heutigen Vielfalt an Hardware, die wir zuhause haben und eingesetzten OS, kann kein Programmierer der Welt alle Fehlerquellen abdecken. Ich kann gut ohne Web-Flashtool leben.
